src/target.rs

```diff
@@ -0,0 +1,164 @@
+        }
+
+        v
+    } else {
+        #[cfg(target_os = "linux")]
+        {
+            linux::detect_targets_linux().await
+        }
+        #[cfg(target_os = "macos")]
+        {
+            macos::detect_targets_macos()
+        }
+        #[cfg(not(any(target_os = "linux", target_os = "macos")))]
+        {
+            from_array([TARGET.into()])
+        }
+    }
+}
+
+// Figure out what the host target is, from rustc or from this program's own build target
+async fn get_targets_from_rustc() -> Option<Box<str>> {
+    match Command::new("rustc").arg("-vV").output().await {
+        Ok(Output { status, stdout, .. }) if status.success() => Cursor::new(stdout)
+            .lines()
+            .filter_map(|line| line.ok())
+            .find_map(|line| {
+                line.strip_prefix("host: ")
+                    .map(|host| host.to_owned().into_boxed_str())
+            }),
+        _ => None,
+    }
+}
+
+fn from_array<T, const LEN: usize, const CAP: usize>(arr: [T; LEN]) -> ArrayVec<T, CAP> {
+    let mut v = ArrayVec::new();
+
+    for elem in arr {
+        v.push(elem);
+    }
+
+    v
+}
+
+#[cfg(target_os = "linux")]
+mod linux {
+    use super::{from_array, ArrayVec, Command, Output, TARGET};
+
+    pub(super) async fn detect_targets_linux() -> ArrayVec<Box<str>, 2> {
+        let abi = parse_abi();
+
+        if let Ok(Output {
+            status: _,
+            stdout,
+            stderr,
+        }) = Command::new("ldd").arg("--version").output().await
+        {
+            let libc_version =
+                if let Some(libc_version) = parse_libc_version_from_ldd_output(&stdout) {
+                    libc_version
+                } else if let Some(libc_version) = parse_libc_version_from_ldd_output(&stderr) {
+                    libc_version
+                } else {
+                    return from_array([create_target_str("musl", abi)]);
+                };
+
+            if libc_version == "gnu" {
+                return from_array([
+                    create_target_str("gnu", abi),
+                    create_target_str("musl", abi),
+                ]);
+            }
+        }
+
+        // Fallback to using musl
+        from_array([create_target_str("musl", abi)])
+    }
+
+    fn parse_libc_version_from_ldd_output(output: &[u8]) -> Option<&'static str> {
+        let s = String::from_utf8_lossy(output);
+        if s.contains("musl libc") {
+            Some("musl")
+        } else if s.contains("GLIBC") {
+            Some("gnu")
+        } else {
+            None
+        }
+    }
+
+    fn parse_abi() -> &'static str {
+        if TARGET.ends_with("abi64") {
+            "abi64"
+        } else if TARGET.ends_with("eabi") {
+            "eabi"
+        } else if TARGET.ends_with("eabihf") {
+            "eabihf"
+        } else if TARGET.ends_with("gnu") || TARGET.ends_with("musl") {
+            ""
+        } else {
+            panic!("Unknown abi")
```

Comment:
I don't think we should panic here. There may be more targets added that don't fit and it would be annoying to start panicking when we didn't before. Also, examples of existing targets that would panic:

- aarch64-unknown-linux-gnu_ilp32
- powerpc-unknown-linux-gnuspe

Reply:
I posted a new commit to fix this by parsing it other way around:

Instead of detecting suffix, we check for the prefix in the last part of the target.

If it starts with `gnu`, we trim `gnu` and return it.
Similarly with `musl`.

The new solution would `panic!` for unrecognizable libc, which I think should be very rare consider majority of them use gnu and some use musl.

There isn't even official support for libc other than glibc and musl, so it should be fine.